src/element/mod.rs

```diff
@@ -578,9 +579,17 @@ impl Element {
     ///
     /// If the input has valid data, returns `Ok(Vec<Element>)`.
     /// If the input has invalid data, returns `Err(IonError)`.
-    pub fn read_all<A: AsRef<[u8]>>(data: A) -> IonResult<Vec<Element>> {
-        let bytes: &[u8] = data.as_ref();
-        ReaderBuilder::default().build(bytes)?.elements().collect()
+    pub fn read_all<A: AsRef<[u8]>>(data: A) -> IonResult<Sequence> {
+        Element::iter(data.as_ref())?.collect()
+    }
+
+    /// Returns an iterator over the Elements in the provided Ion data source.
+    /// If the data source cannot be read or contains invalid Ion data, this method
+    /// will return an `Err`.
+    pub fn iter<'a, I: ToIonDataSource + 'a>(
+        source: I,
+    ) -> IonResult<impl Iterator<Item = IonResult<Element>> + 'a> {
+        Ok(ReaderBuilder::default().build(source)?.into_elements())
```

Comment:
🗺️ For the 1.0 stabilization process, we plan to label the streaming APIs as experimental and make them opt-in. This method will allow folks who aren't using `Reader`/`ReaderBuilder` to iterate over the elements in a data source without reading the whole thing into memory first.